File: Spark operations (e.g., transformations and actions) and MapReduce/Scala/task2.scala
import org.apache.spark.sql.SparkSession
import java.io._
import org.json4s
import scala.collection.mutable
import scala.io.Source
import org.json4s._
import org.json4s.JsonDSL._
import org.json4s.jackson.JsonMethods._


object task2 {

  def main(args: Array[String]): Unit ={

    if (args.length != 5) {
      println("Please enter all the required agruments")
      sys.exit(1)
    }

    val ss = SparkSession
      .builder()
      .appName(name = "task2")
      .config("spark.master", "local[*]")
      .getOrCreate()

    val sc = ss.sparkContext
    sc.setLogLevel("ERROR")

    if (args(3) == "spark"){

      val review_rdd = sc.textFile(args(0).toString).map(string => parse(string).values.asInstanceOf[Map[String,String]])

      val business_rdd = sc.textFile(args(1).toString).map(string => parse(string).values.asInstanceOf[Map[String,String]])

      val b =business_rdd.filter(x=> x("categories") != null)
        .flatMap(x=> for (i <- x("categories").split(",")) yield (x("business_id"), i.trim))

      val r =review_rdd.map(x=>(x("business_id"), x("stars")))

      val s= r.join(b)

      val top_categories = s.map(x => (x._2._2, x._2._1)).map(x => (x._1, (x._2.asInstanceOf[Double], 1)))
        .reduceByKey((x, y) => {
          (y._1.asInstanceOf[Double] + x._1.asInstanceOf[Double], x._2 + y._2)
        }).mapValues{ case (sum , count) => sum/count}.sortBy(x => (-x._2,x._1)).take(args(4).toInt)

      val output_file = new File(args(2))
      val output_writer = new BufferedWriter(new FileWriter(output_file))
      output_writer.write("{"+ "\"result\":"+top_categories.toList.toString().replaceAll("List","").replaceAll("\\(", "[").replaceAll("\\)", "]")+"}")
      output_writer.close()

    }

    if (args(3) == "no_spark"){
      val review_map:mutable.Map[String,Tuple2[Double, Int]] = mutable.Map()
      val json_values = Source.fromFile(args(0))("UTF-8")
      for(line <- json_values.getLines()) {
        val json_line = parse(line)
        val b = (json_line \ "business_id").asInstanceOf[json4s.JString].s
        val stars = (json_line \ "stars").asInstanceOf[_root_.org.json4s.JsonAST.JDouble].num
        if(!review_map.contains(b)) {
          review_map += b -> (stars,1)
        }
        else {
          val t = review_map(b)
          val sum  = t._1 + stars
          val count = t._2 + 1
          review_map(b) = (sum, count)
        }
      }

      val categories_map:mutable.Map[String,Tuple2[Double, Int]] = mutable.Map()
      val categories_values = Source.fromFile(args(1))("UTF-8")
      for(lines <- categories_values.getLines()) {
        val line = parse(lines)
        val b = (line \ "business_id").asInstanceOf[json4s.JString].s
        val category = (line \ "categories")
        var categories:String = null

        if (category.values!= null) {
          categories = category.asInstanceOf[json4s.JString].s
        }
        if(categories != null && categories.nonEmpty) {
          val categories_array= categories.split(",")
          for (cat <- categories_array) {
            if(!categories_map.contains(cat.trim)){
              if(review_map.contains(b)){
                categories_map(cat.trim) = (review_map(b)._1, review_map(b)._2)
              }
            }
            else {
              val v = categories_map(cat.trim)
              if(review_map.contains(b)){
                val rev = review_map(b)
                val s = v._1 + rev._1
                val c = v._2 + rev._2
                categories_map(cat.trim) = (s,c)
              }
            }
          }
        }
      }

      val final_map:mutable.Map[String,Double] = mutable.Map()
      for((k, v) <- categories_map) {
        final_map(k) = v._1/v._2
      }
      val sort_map = final_map.toList.sortBy(r=>(-r._2, r._1))

      val json : JObject = "result" -> sort_map.slice(0,args(4).toInt)

      val output_file = new File(args(2))
      val output_writer = new BufferedWriter(new FileWriter(output_file))
      val v =pretty(render(json)).replaceAll("\\{", "[").replaceAll("\\}", "]") +"}"
      output_writer.write(v.replaceFirst("\\[", "{").replaceAll("\\]\\}", "}"))
      output_writer.close()

    }

    sc.stop()
  }

}

File: Spark operations (e.g., transformations and actions) and MapReduce/Python/task2.py
from pyspark import SparkContext
import json
import sys


def sub_tasks(review, business, output_file, if_spark, n):
    if str(if_spark) == "spark":
        sc = SparkContext('local[*]', 'task2')

        review_rdd = sc.textFile(str(review)).map(lambda x: json.loads(x)).cache()
        business_rdd = sc.textFile(str(business)).map(lambda x: json.loads(x)).cache()

        b = business_rdd.filter(lambda x: x['categories'] is not None) \
            .flatMap(lambda x: [(x['business_id'], i.strip()) for i in x['categories'].split(',')]).cache()

        r = review_rdd.map(lambda x: (x['business_id'], x['stars'])).cache()

        s = r.join(b)

        top_categories = s.map(lambda x: (x[1][1], x[1][0])).groupByKey().mapValues(lambda x: sum(x) / len(x)) \
            .sortBy(lambda x: (-x[1], x[0])).take(int(n))

        output = {
            "result": top_categories
        }

        with open(str(output_file), "w+") as fileout:
            json.dump(output, fileout, indent=4)

        sc.stop()

    elif str(if_spark) == "no_spark":
        review_dict = {}
        with open(str(review)) as f:
            for l in f:
                obj = json.loads(l)
                id = str(obj['business_id'])
                stars = obj['stars']
                if id not in review_dict:
                    review_dict[id] = (stars, 1)
                else:
                    val = review_dict.get(id)
                    s = val[0] + stars
                    count = val[1] + 1
                    review_dict[id] = (s, count)

        category_dict = {}
        with open(str(business), encoding="utf8") as f:
            for l in f:
                obj = json.loads(l)
                id = str(obj['business_id'])
                if obj['categories'] is not None:
                    category_list = obj['categories'].split(',')
                    for cat in category_list:
                        cat = cat.strip()
                        if cat not in category_dict:
                            if id in review_dict:
                                category_dict[cat] = review_dict[id]
                        else:
                            val = category_dict.get(cat)
                            rev = review_dict.get(id)
                            if rev is not None and rev[0] is not None and rev[1] is not None:
                                s = val[0] + rev[0]
                                count = val[1] + rev[1]
                                category_dict[cat] = (s, count)

        for key, value in category_dict.items():
            category_dict[key] = value[0] / value[1]

        top_categories = sorted(category_dict.items(), reverse=False, key=lambda x: (-x[1], x[0]))

        output = {
            "result": top_categories[:int(n)]
        }

        with open(str(output_file), "w+") as fileout:
            json.dump(output, fileout, indent=4)


def main():
    if len(sys.argv) != 6:
        print("Please enter all the required agruments")
        exit(-1)

    sub_tasks(sys.argv[1], sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])


if __name__ == "__main__": main()
